urldee901fae6_DocX_personal_timetabling_tgz-pJ8-net_personaltt_main_BenchmarkJ8/src/net/personaltt/solver/heuristics/MinDurationConflictAllocationCost.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package net.personaltt.solver.heuristics;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import net.personaltt.model.Occurrence;
+import net.personaltt.solver.core.SolverState;
+import net.personaltt.utils.BaseInterval;
+import net.personaltt.utils.ValuedInterval;
+import net.personaltt.utils.intervalmultimap.IntervalMultimap;
+
+/**
+ * Minimal duration conflict cost of allocation.
+ * Cost is sum of overlapping lengths of other occurrences which are weighted by
+ * if it is length above minimal duration. Lengths that fits into over minimal
+ * duration are halved in cost.
+ * @author docx
+ */
+public class MinDurationConflictAllocationCost extends OccurrenceConflictAllocationCost {
+
+
+    
+    public MinDurationConflictAllocationCost(Occurrence allocatingOccurrence, SolverState solution)  {
+        this.allocatingOccurrence = allocatingOccurrence;
+        this.solution = solution;
+    }
+    
+    @Override
+    public int occurrenceOverMinDurationCost(Occurrence occurrence, BaseInterval<Integer> allocation) {
+        if (occurrence.equals(this.allocatingOccurrence)) {
+            return 0;
+        }
+         
+        // for all occurrences that are overlaping start point, count duration
+        // that is cropped by its allocation
+        int durationInOccurrence = 
+                Math.min(
+                    occurrence.getAllocation().getStart() + occurrence.getAllocation().getDuration(), 
+                    allocation.getEnd()) - 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;allocation&#39; (index: 1 | type: BaseInterval). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;occurrence&#39;. The same parameter type is validated here for example: urldee901fae6_DocX_personal_timetabling_tgz-pJ8-net_personaltt_main_BenchmarkJ8/src/net/personaltt/solver/heuristics/ConflictSumAllocationCost.java:48. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).